src/backend/wayland/application.rs

```diff
@@ -14,582 +14,222 @@
-                                surfaces::Outputs::inserted(&win, &output);
-                            }
-                        }
-                        outputs::Event::Removed(output) => {
-                            tracing::debug!("output removed {:?} {:?}", output.gid, output.id());
-                            appdata.outputs.borrow_mut().remove(&output.id());
-                            for (_, win) in appdata.handles_iter() {
-                                surfaces::Outputs::removed(&win, &output);
-                            }
+        let mut state = self
+            .state
+            .borrow_mut()
+            .take()
+            .expect("Can only run an application once");
+        state.handler = handler;
+        let mut event_loop = state.event_loop.take().unwrap();
+        event_loop
+            .run(None, &mut state, |state| loop {
+                match state.idle_actions.try_recv() {
+                    Ok(IdleAction::Callback(cb)) => cb(state),
+                    Ok(IdleAction::Token(window, token)) => match state.windows.get_mut(&window) {
+                        Some(state) => state.handler.idle(token),
+                        None => {
+                            tracing::debug!("Tried to run an idle token on a non-existant window")
                         }
                     },
+                    Err(TryRecvError::Empty) => break,
+                    Err(TryRecvError::Disconnected) => {
+                        unreachable!("Backend has allowed the idle sender to be dropped")
+                    }
                 }
             })
-            .unwrap();
-
-        handle
-            .insert_source(timer_source, move |token, _metadata, appdata| {
-                tracing::trace!("timer source {:?}", token);
-                appdata.handle_timer_event(token);
-            })
-            .unwrap();
-
-        let signal = eventloop.get_signal();
-        let handle = handle.clone();
-
-        let res = eventloop.run(Duration::from_millis(20), &mut self.data, move |appdata| {
-            if appdata.shutdown.get() {
-                tracing::debug!("shutting down, requested");
-                signal.stop();
-                return;
-            }
-
-            if appdata.handles.borrow().len() == 0 {
-                tracing::debug!("shutting down, no window remaining");
-                signal.stop();
-                return;
-            }
-
-            Data::idle_repaint(handle.clone());
-        });
-
-        match res {
-            Ok(_) => tracing::info!("wayland event loop completed"),
-            Err(cause) => tracing::error!("wayland event loop failed {:?}", cause),
-        }
+            .expect("Shouldn't error in event loop");
     }
 
     pub fn quit(&self) {
-        self.data.shutdown.set(true);
+        // Stopping the event loop should be sufficient, as our state is dropped upon `run` finishing
+        self.loop_signal.stop();
```

Comment:
I'd actually wake the event loop from the signal as well since you may need to wait for some more events to come in if this is done just before the event loop suspends the process.